src/controller/CHIPDeviceController.h

```diff
@@ -616,10 +616,15 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController, public SessionEst
      *   This function processes the CSR sent by the device.
      *   (Reference: Specifications section 11.22.5.8. OpCSR Elements)
      *
-     * @param[in] device The object that contains device's current state.
-     * @param[in] CSR    The Certificate Signing Request.
+     * @param[in] CSR             The Certificate Signing Request.
+     * @param[in] CSRNonce        The Nonce sent by us when we requested the CSR.
+     * @param[in] VendorReserved1 vendor-specific information that may aid in device commissioning.
+     * @param[in] VendorReserved2 vendor-specific information that may aid in device commissioning.
+     * @param[in] VendorReserved3 vendor-specific information that may aid in device commissioning.
+     * @param[in] Signature       Cryptographic signature generated for all the above fields.
      */
-    CHIP_ERROR ProcessOpCSR(Device * device, const ByteSpan & CSR);
+    CHIP_ERROR ProcessOpCSR(const ByteSpan & CSR, const ByteSpan & CSRNonce, const ByteSpan & VendorReserved1,
+                            const ByteSpan & VendorReserved2, const ByteSpan & VendorReserved3, const ByteSpan & Signature);
 
```

Comment:
A more scalable approach is to have:

- The entire NOCSRElements as a ByteSpan
- The AttestationSignature as a ByteSpan
- A placeholder argument for 3 bytespans: DAC cert, PAI cert, PAA cert

See spec:

```
[width="99%",frame="all",options="header", cols="^,<s,<,^,^"]
|===
| Id | Field                   | Type      | Constraint                       | Conformance
| 0  | NOCSRElements           | octstr    | max <<ref_RESP_MAX,`RESP_MAX`>>  | M
| 1  | AttestationSignature    | octstr    | 64                               | M
|===
```

That would allow:

- The full payload to be passed to a CA
- The CSR portion of the `NOCSRElements` to be extracted
- Reducing the "breadth" of this public API that needs to often pass data to code that lives outside native C++ code.

The main change would be to go from "pre-processing" the inner TLV of `NOCSRElements` to post-processing it. In the synchronous call path of this command handling, it would take no additional code, just require moving the TLV decode a bit deeper in the chain.

I recommend this now because the to-spec processing of CSRResponse requires device attestation validation (which is in the works), and the public API currently makes this impossible. Since the API is being touched, it's a good time to do a little bit of reshuffling that opens the door to doing attestation signature validation on the CSR and also opening the door to end-to-end attestation of CSRReponse by CA of the commissioner (e.g. if the CA code lives in a trustzone or in a cloud, rather than in-process as is the case of the examples today).

Reply:
@pan-apple - PRs new require all conversations to be resolved before merging. Not sure how to interpret the thumbs up: will you follow up with a commit with additional changes?

Reply:
Sorry, I wanted to address the review comment as part of the PR, but other PRs kept taking my time.
@tcarmelveilleux would it be ok to tackle the suggestion as a separate PR, as I cannot get back to updating this part of the code in next few days.